Navigation: Add Post, Page, Category and Tag variations to Link

noisysocks opened the pull request:
Closes #22919. ~Requires #22600.~ Split out from #24613.

![nav-variations](https://user-images.githubusercontent.com/612155/90710862-a19b0980-e2e2-11ea-9442-f6a0714bc7b3.gif)

Adds Post, Page, Category and Tag variations to the Navigation Link block. Each variation sets the `type` attribute which in turn causes `LinkControl` to filter its results using the `/wp/v2/search` endpoint's `type` and `subtype` params.

~This PR requires #22600 which adds the ability to search for categories and triages using the search endpoint. Do not merge this PR until #22600 is merged and this PR's base is updated.~

### How to test

1. Ensure that your site has some posts, pages, categories and tags.
1. Create a Navigation block.
1. Add a Link block. You should be able to search for posts, pages, categories and tags.
1. Add a Page block. You should only be able to search for posts.
1. ...and so on for Post, Category and Tag.

noisysocks committed:
Navigation: Add Post, Page, Category and Tag variations to Link

Adds Post, Page, Category and Tag variations to the Navigation Link
block. Each variation sets the type attribute which in turn causes
LinkControl to filter its results using the /wp/v2/search API's type and
subtype params.

noisysocks commented:
@shaunandrews: Does this flow look right to you?

talldan commented on packages/editor/src/components/provider/index.js:
```diff
@@ -37,36 +37,59 @@ import ConvertToGroupButtons from '../convert-to-group-buttons';
+	}
 
-	return Promise.all( [ posts, terms, formats ] ).then( ( results ) => {
+	return Promise.all( queries ).then( ( results ) => {
```
Not introduced by this PR, but this does result in all of the results being smooshed together in the suggestions in a possibly unintended way.

Posts are always first, followed by terms and then post formats, even if the latter is more relevant to the search query.

Maybe this needs to be a REST API change for it work properly. Alternatively, if the Link block only handled external links it wouldn't be as much of an issue.

noisysocks replied:
Hm, tough one.

A related problem is that `per_page` doesn't really work because if we set e.g. `per_page = 3` we'll end up with 9 results because we're performing 3 requests.

I'm not really sure how to fix this. Right now the REST API endpoint is architected in a way such that `type` is required and can only be set to one value. Perhaps that needs to change? cc. @TimothyBJacobs 

> Posts are always first, followed by terms and then post formats, even if the latter is more relevant to the search query.

Note that `/wp/v2/search` doesn't order its results by relevancy. But, yes, I get what you mean.

TimothyBJacobs replied:
We could change the search endpoint to accept multiple types, but that doesn't really solve the problem as far as I can see. Each search handler is distinct, so we can't do one sql query to combine them all. So we'll wind up with the same pagination issue. We can divide the `per_page` by the number of search handlers, but if a search handler doesn't fill its entire "slot" then we'll return an inadequate number of responses.

noisysocks replied:
In the upstream branch I made it so that it'll fetch `perPage * 3` results and then slice the array to have only `perPage` entries. This means posts take priority over terms and post formats, which I think is right.

talldan reviewed (commented):
This is actually a surprisingly straightforward change, which is nice.

In terms of user experience, I think it may be confusing that there's an overlap between the block functionality. I wonder if we should limit 'Link' to just being external links, 'Page' to just being pages etc.

I personally don't expect to get other options when I add those blocks, especially the 'Link' block which gets suggestions for everything.

noisysocks commented on packages/block-editor/src/components/link-control/use-search-handler.js:
```diff
@@ -117,12 +120,12 @@ export default function useSearchHandler(
 				: handleEntitySearch(
 						val,
-						args,
+						{ isInitialSuggestions, type, subtype },
```
Good flag. How's ed791581c6 look to you? I made `LinkControl` pass along whatever `suggestionsQuery` it gets plus an additional `isInitialSuggestions`. It's then up to the `fetchLinkSuggestions` function defined in `editor` to interpret that into something compatible with the REST API.

talldan replied:
Looks good 👍

noisysocks commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -109,6 +109,7 @@ function LinkControl( {
 	createSuggestion,
 	withCreateSuggestion,
 	inputValue: propInputValue = '',
+	suggestionsQuery = {},
```
👍 64142c4d34

noisysocks commented on packages/editor/src/components/provider/index.js:
```diff
@@ -37,36 +37,59 @@ import ConvertToGroupButtons from '../convert-to-group-buttons';
-			type: 'post',
-		} ),
-	} );
+const fetchLinkSuggestions = async (
```
Yes, whoops. Fixed in b4a1023b58.

noisysocks commented:
> In terms of user experience, I think it may be confusing that there's an overlap between the block functionality. I wonder if we should limit 'Link' to just being external links, 'Page' to just being pages etc.

I'll defer to @shaunandrews on this one 😀

TimothyBJacobs commented:
Why does this have a hardcoded list of post types and taxonomies?

noisysocks commented:
This has some test failures I need to fix up.

noisysocks commented:
> Why does this have a hardcoded list of post types and taxonomies?

I started by implementing the most common Link variations: Page, Post, Category, Tag. You're probably right though that this should be dynamic so that we have a Link block variation for every (custom) post type and (custom) taxonomy.

TimothyBJacobs commented:
> > Why does this have a hardcoded list of post types and taxonomies?
> 
> I started by implementing the most common Link variations: Page, Post, Category, Tag. You're probably right though that this should be dynamic so that we have a Link block variation for every (custom) post type and (custom) taxonomy.

:+1: Yeah I think that'd be a worthwhile change. Would also make sure any changes to the post type label are picked up.

shaunandrews commented:
Very cool to see this; I like having a block variation for each post type — though I worry a little about how this scales. Does a site with 20+ post type's defined then list 20+ link variations? If so, is that a problem?

I'd expect a Page Link block to only list pages in it's default suggests and search; and similar, a Post Link block would only show posts.

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/191598/91223828-3640a580-e6ef-11ea-8729-0520c152a721.png">

The more general _catchall_ Link block should still be able to do it all, allowing the display of any post type along with supporting general URL links.

I tried creating some new icons that use the Navigation icon in an attempt to visually _brand_ these blocks together, but I think using the existing, more generic icons might work just as good, it not better. The new icons can be found in the [WP Design Library Figma](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=1944%3A2472):

<img width="828" alt="image" src="https://user-images.githubusercontent.com/191598/91223380-89662880-e6ee-11ea-99b0-0eb48c80e69d.png">

TimothyBJacobs commented:
> Very cool to see this; I like having a block variation for each post type — though I worry a little about how this scales. Does a site with 20+ post type's defined then list 20+ link variations? If so, is that a problem?

Yeah this can get annoying in Core currently. For example if you have a site with a lot of CPTs/Taxonomies and open the Menus screen you wind up with a bunch of meta boxes and the screen can get a bit unwieldy. You can hide them via Screen Options, and I think(?) only some are shown by default which can be really confusing for the user.

Making it a variation, though, feels like it might be a somewhat better experience? Since you are first searching & selecting the type and then that visual noise goes away.

noisysocks committed:
Navigation: Add icons to each of the Link variations

noisysocks committed:
Navigation: Show 'Create post' or 'Create page' when appropriate

- Display 'Create post' when inserting a Post Link.
- Display 'Create page' when inserting a Link or a Page Link.
- Only allow direct URL input when inserting a Link.

noisysocks committed:
Navigation: Display 'post_tag' as 'tag'

noisysocks committed:
Fix LinkControl unit tests and Navigation E2E tests

noisysocks committed:
LinkControl: Add tests for noURLSuggestion and createSuggestionButtonText

noisysocks commented:
This is ready for a re-review.

I rebased the PR to include https://github.com/WordPress/gutenberg/pull/22600 and fixed all of the test failures.

I added the basic icons to each variation per @shaunandrews's suggestion above.

![Screen Shot 2020-08-26 at 16 01 54](https://user-images.githubusercontent.com/612155/91267312-74ed6300-e7b6-11ea-8b4e-3eaa41ae686c.png)

"Create page" will appear at the bottom of the Link and Page variations and "Create post" will appear at the bottom of the Post variation. No creation is possible in the Tag and Category variations. Linking to a user inputted URL is only possible in the Link variation.

| Link | Page | Post | Tag | Category |
| - | - | - | - | - |
| <img width="396" alt="Screen Shot 2020-08-26 at 16 03 43" src="https://user-images.githubusercontent.com/612155/91267444-bbdb5880-e7b6-11ea-8edb-f6b27502d486.png">| <img width="384" alt="Screen Shot 2020-08-26 at 16 03 59" src="https://user-images.githubusercontent.com/612155/91267455-c1d13980-e7b6-11ea-86ef-7dd0ec00799a.png"> | <img width="394" alt="Screen Shot 2020-08-26 at 16 03 30" src="https://user-images.githubusercontent.com/612155/91267464-c695ed80-e7b6-11ea-84ff-ce25f5104213.png"> | <img width="389" alt="Screen Shot 2020-08-26 at 16 04 18" src="https://user-images.githubusercontent.com/612155/91267471-ca297480-e7b6-11ea-89de-7192b9f14457.png"> | <img width="379" alt="Screen Shot 2020-08-26 at 16 04 31" src="https://user-images.githubusercontent.com/612155/91267476-ce559200-e7b6-11ea-8819-47e230842907.png"> |

I agree that a variation should be added for each custom post type and taxonomy, but I would like to do this later as it's a little bit tricky to implement.

draganescu commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -298,8 +334,29 @@ function NavigationLinkEdit( {
+									}
+									return createInterpolateElement(
+										sprintf( format, searchTerm ),
+										{ mark: <mark /> }
```
So nice 👍

draganescu reviewed (approved):
Code looks good and works like a charm! :shipit:

noisysocks committed:
LinkControl: Improve JSDoc comment

noisysocks commented:
> I agree that a variation should be added for each custom post type and taxonomy, but I would like to do this later as it's a little bit tricky to implement.

Issue here: https://github.com/WordPress/gutenberg/issues/24814

shaunandrews commented:
At second glance, I don't think we need to post_type label in the search results if its only listing a single post_type.

noisysocks commented:
> At second glance, I don't think we need to post_type label in the search results if its only listing a single post_type.

Issue here: https://github.com/WordPress/gutenberg/issues/24839
